Add flags option to table loading

flxo opened the pull request:
When loading a verity table the device *must* be a readonly. The dm driver requires the flags for the `DM_TABLE_LOAD_CMD` ioctl to include at least `DM_READONLY` - otherwise it complains:

```
May 25 19:32:42 XXX kernel: [ 7763.805044] device-mapper: table: 254:7: verity: Device must be readonly
May 25 19:32:42 XXX kernel: [ 7763.805049] device-mapper: ioctl: error adding target to table
```

Add `core::dm::table_load_flags` that enables any user to pass `DmFlags::DM_READONLY` to the `DM_TABLE_LOAD_CMD` operation.

An Option could be to extend `table_load` with an `Option<DmFlags>`. This changes the API in a breaking way, so I decided to add `table_load_flags`. Let me know if you prefer the `Option` argument.

It's possible that other options than `DM_READONLY` are allowed for the `DM_TABLE_LOAD`. Would be easy to add to `allowed_flags`.

mulkieran commented:
Probably blocked by #624

flxo commented:
> Probably blocked by #624

Shall I rebase this PR onto #624 (https://github.com/jbaublitz/devicemapper-rs/tree/cleanup)?

mulkieran commented:
> > Probably blocked by #624
> 
> Shall I rebase this PR onto #624 (https://github.com/jbaublitz/devicemapper-rs/tree/cleanup)?

Probably not until next week. We can ping you when #624 is in.

mulkieran commented:
@flxo ping! We got the changes in; sorry for the delay. Please resolve conflicts, etc., and we'll begin review.

flxo commented:
@mulkieran Thanks. I'll rebase tomorrow.

flxo commented:
@mulkieran I rebased the PR. 

Apropos tests: I'm unsure if a test for `table_load` and `table_load_flags` is applicable because this would need a backing device. Maybe you have a hint regarding this.

mulkieran commented:
@flxo Thanks for the PR! Is your ultimate goal for devicemapper-rs to support dm-verity?

flxo commented:
> @flxo Thanks for the PR! Is your ultimate goal for devicemapper-rs to support dm-verity?

Exactly. I borrowed some code from `devicemapper-rs` a year ago and slightly patched it to be be able to setup verity tables - mainly adding the `ro` flag. The project is [Northstar](https://github.com/esrlabs/northstar). Experimental "containers" with squashes images that are checked via verity.
 
 I thought it's time to upstream and cleanup the patch and get rid of this "fork".

mulkieran commented:
@flxo Do you anticipate wanting to add a lot more changes to devicemapper-rs? Because devicemapper-rs is major version 0, we feel free to go with the alternative for this PR you suggested, the API-breaking approach, in the interests of better maintainability in future. Since it does break the API, though, we'll want to cut a new release of devicemapper-rs first, to include the non-breaking changes that we already have merged.

So we'ld like to know if your plans are likely to involve more changes to devicemapper-rs than you've already put up. If so, it might even be best to set up a meeting, so we can arrange to coordinate better. Let us know. Thanks.

flxo commented:
@mulkieran I've no further features requests or PRs in the pipeline. The ability to add the `ro` flag when loading a table is the only thing that I missed for my purpose. I'm already using my PR branch for testing.

I do not fully understand "we feel free to go with the alternative for this PR". Which alternative?

Thanks for your review!

flxo commented:
@mulkieran I just noticed that the merge of #633 broken the build for target `aarch64-linux-android`. I'll propose a separate PR similar to #626. This will not include any API change.

mulkieran commented:
Blocked by #640

mulkieran commented:
> @mulkieran I just noticed that the merge of #633 broken the build for target `aarch64-linux-android`. I'll propose a separate PR similar to #626. This will not include any API change.

Thanks! We would very much like to add some CI to catch these kinds of errors in future...but this part is proving not unexpectedly challenging. We may still overcome the difficulties, however. In the meantime, though, feel free to put up a fix to #640 whenever is convenient for you.

mulkieran commented:
> I do not fully understand "we feel free to go with the alternative for this PR". Which alternative?

@flxo We think that 

"An Option could be to extend table_load with an Option<DmFlags>. This changes the API in a breaking way, so I decided to add table_load_flags. Let me know if you prefer the Option argument."

would be a better idea probably. But we would want to cut a new release firstt. If so, we'll expect to have that done by the end of this week.

flxo commented:
@mulkieran Thanks for you reply. I'll update the PR with the flags as an `Option<DmFlags>` e.g:

```rust
pub fn table_load(&self, id: &DevId, targets: &[(u64, u64, String, String)], flags: Option<DmFlags>) -> DmResult<DeviceInfo>
```

mulkieran commented:
Great, thx. Blocked by #654 .

mulkieran commented:
@jbaublitz I'm inclined to think we should be more rather than less invasive here. I just cut the new release of devicemapper-rs; so I think we should feel comfortable with just changing the signatures of both ```table_load``` methods, the one in shared.rs and the one in dm.rs. Let me know what you think.

@flxo Would you be just as happy w/ that?

flxo commented:
> @flxo Would you be just as happy w/ that?

Sure. Whatever you prefer. My personal preference is the `Option<Flags>` variant over the `fn table_load_flags(...)` version.

jbaublitz reviewed (approved):
@mulkieran Okay that sounds good to me! After reviewing the code, I agree.

flxo commented:
Oops. I just noticed that I didn't push the modification with the `Option` arg at all. It's here: https://github.com/flxo/devicemapper-rs/tree/pr-table-ro-option-arg

mulkieran commented:
> Oops. I just noticed that I didn't push the modification with the `Option` arg at all. It's here: https://github.com/flxo/devicemapper-rs/tree/pr-table-ro-option-arg

I believe we would all prefer that. Pls apply to this PR. Thx!

flxo commented:
> > Oops. I just noticed that I didn't push the modification with the `Option` arg at all. It's here: https://github.com/flxo/devicemapper-rs/tree/pr-table-ro-option-arg
> 
> I believe we would all prefer that. Pls apply to this PR. Thx!

Pushed.

mulkieran commented on src/core/dm.rs:
```diff
@@ -497,7 +502,9 @@ impl DM {
         }
 
-        let mut hdr = DmOptions::new().to_ioctl_hdr(Some(id), DmFlags::empty())?;
+        let mut hdr = DmOptions::new()
```
Please look at, e.g., ```remove_all``` in this file to see how a method that accepts flags is typically handled. The key point is that the method should take the DmOptions, not the DmFlags, for consistency.

It seems that the second argument to ```to_ioctl_hdr``` is the allowed flags, i.e. that value gets anded with the flags actually passed.

mulkieran reviewed (changes requested):
One request, I think otherwise it should be fine.

flxo commented:
I don't know how to deal with the CI error on `Fedora`:

```
Run COMPARE_FEDORA_VERSIONS=./compare_fedora_versions MANIFEST_PATH=../../Cargo.toml FEDORA_RELEASE=rawhide make -f ../../Makefile check-fedora-versions
echo "Testing that COMPARE_FEDORA_VERSIONS environment variable is set to a valid path"
Testing that COMPARE_FEDORA_VERSIONS environment variable is set to a valid path
test -e "./compare_fedora_versions"
./compare_fedora_versions --manifest-path=../../Cargo.toml --release=rawhide
Version 1.0.3 of crate semver is available in Fedora. Requires update in Cargo.toml
{"missing": [], "high": [], "low": ["semver"]}
make: *** [../../Makefile:65: check-fedora-versions] Error 1
Error: Process completed with exit code 2.
```

Can this be related to this patch?

mulkieran commented:
> I don't know how to deal with the CI error on `Fedora`:
> 
> ```
> Run COMPARE_FEDORA_VERSIONS=./compare_fedora_versions MANIFEST_PATH=../../Cargo.toml FEDORA_RELEASE=rawhide make -f ../../Makefile check-fedora-versions
> echo "Testing that COMPARE_FEDORA_VERSIONS environment variable is set to a valid path"
> Testing that COMPARE_FEDORA_VERSIONS environment variable is set to a valid path
> test -e "./compare_fedora_versions"
> ./compare_fedora_versions --manifest-path=../../Cargo.toml --release=rawhide
> Version 1.0.3 of crate semver is available in Fedora. Requires update in Cargo.toml
> {"missing": [], "high": [], "low": ["semver"]}
> make: *** [../../Makefile:65: check-fedora-versions] Error 1
> Error: Process completed with exit code 2.
> ```
> 
> Can this be related to this patch?

Yes and no. semver crate was just bumped a major version in Fedora, which triggered the failure. That particular check is meant to catch if a PR adds a new dependency not available in Fedora rather than this version bump.

Please rebase and all checks should pass.

mulkieran reviewed (changes requested):
@flxo Whoops! We realized in a meeting that we would like to request one last change. Can you make the ```table_load``` ```options``` parameter not an Option? This makes it more consistent w/ the rest of the methods and leaves only one way to indicate "no options requested". With that change, I should be able to merge immediately.

flxo commented:
> @flxo Whoops! We realized in a meeting that we would like to request one last change. Can you make the `table_load` `options` parameter not an Option? This makes it more consistent w/ the rest of the methods and leaves only one way to indicate "no options requested". With that change, I should be able to merge immediately.

Sure! Updated as requested. I also fixed 2 typos ;-)

mulkieran commented:
Nearly done. I think this is a real compiler error now.

flxo commented:
> Nearly done. I think this is a real compiler error now.

Sorry. My bad. I fixed the doc compile error.

mulkieran commented:
thx!
